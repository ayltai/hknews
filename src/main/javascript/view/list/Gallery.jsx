import PropTypes from 'prop-types';
import React from 'react';
import { SingleFieldList, } from 'react-admin';
import Lightbox from 'react-image-lightbox';
import 'react-image-lightbox/style.css';

import { StringUtils, } from '../../util/StringUtils';

const getEntries = props => props.ids && props.data ? props.ids.map(id => props.data[id]) : [];

const getImages = props => Object.entries(getEntries(props)).map(record => ({
    label  : StringUtils.decode(record[1][props.caption]),
    source : record[1][props.source],
}));

export const Gallery = props => {
    const [ index, setIndex, ]      = React.useState(0);
    const [ open,  setOpen,  ]      = React.useState(false);
    const { source, ...otherProps } = props;
    const entries                   = getEntries(props);
    const images                    = getImages(props);

    const children = React.Children.map(props.children, child => React.cloneElement(child, {
        onClick : event => {
            setIndex(Math.max(0, entries ? entries.findIndex(entry => entry[source] === event.target.src) : 0));
            setOpen(true);
        },
    }));

    return (
        <React.Fragment>
            <SingleFieldList
                linkType={false}
                {...otherProps}>
                {children[0]}
            </SingleFieldList>
            {open && (
                <Lightbox
                    mainSrc={images[index].source}
                    nextSrc={index + 1 < images.length ? images[(index + 1) % images.length].source : undefined}
                    prevSrc={index - 1 >= 0 ? images[(index + images.length - 1) % images.length].source : undefined}
                    imageCaption={images[index].label}
                    onCloseRequest={() => setOpen(false)}
                    onMoveNextRequest={() => setIndex((index + 1) % images.length)}
                    onMovePrevRequest={() => setIndex((index + images.length - 1) % images.length)} />
            )}
        </React.Fragment>
    );
};

Gallery.propTypes = {
    ids      : PropTypes.arrayOf(PropTypes.number),
    data     : PropTypes.object,
    caption  : PropTypes.string,
    source   : PropTypes.string,
    children : PropTypes.object,
};
