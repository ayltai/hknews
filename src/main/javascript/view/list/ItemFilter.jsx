import React from 'react';
import { Filter, ReferenceInput, SelectInput, } from 'react-admin';

export const ItemFilter = props => (
    <Filter {...props}>
        <ReferenceInput
            label='labels.publisher'
            source='source'
            reference='sources'
            allowEmpty
            alwaysOn>
            <SelectInput
                optionText='name'
                variant='standard' />
        </ReferenceInput>
    </Filter>
);
