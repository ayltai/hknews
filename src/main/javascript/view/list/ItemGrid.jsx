import { Card, CardActions, CardActionArea, CardContent, CardHeader, CardMedia, Grid, makeStyles, Tooltip, } from '@material-ui/core';
import PropTypes from 'prop-types';
import React from 'react';
import { RichTextField, ShowButton, TextField, } from 'react-admin';
import TimeAgo from 'react-timeago';

import { StringUtils, } from '../../util/StringUtils';
import { Constants, } from '../../Constants';

export const ItemGrid = ({ basePath, ids, data, }) => {
    const classes = makeStyles(theme => ({
        container   : {
            padding  : theme.spacing(2),
            flexGrow : 1,
        },
        card        : {
            display : 'flex',
        },
        details     : {
            display       : 'flex',
            flexDirection : 'column',
            flexGrow      : 1,
        },
        content     : {
            flex : '1 0 auto',
        },
        title       : {
            display         : '-webkit-box',
            overflow        : 'hidden',
            fontSize        : '120%',
            fontWeight      : 'bold',
            textOverflow    : 'ellipsis',
            whiteSpace      : 'break-spaces',
            WebkitBoxOrient : 'vertical',
            WebkitLineClamp : 1,
        },
        description : {
            minHeight       : theme.spacing(12),
            display         : '-webkit-box',
            overflow        : 'hidden',
            textOverflow    : 'ellipsis',
            WebkitBoxOrient : 'vertical',
            WebkitLineClamp : 4,
        },
        media       : {
            minWidth   : theme.spacing(20),
            width      : theme.spacing(20),
            minHeight  : theme.spacing(20),
            height     : theme.spacing(20),
            marginLeft : theme.spacing(2),
        },
        actions     : {
            display : 'block',
        },
    }))();

    return (
        <div className={classes.container}>
            <Grid
                container
                spacing={2}>
                {ids.map(id => (
                    <Grid
                        item
                        key={id}
                        lg={12}
                        xl={6}>
                        <Card>
                            <CardActionArea href={`#${basePath}/${id}/show`}>
                                <CardHeader
                                    className={classes.content}
                                    title={
                                        <TextField
                                            record={data[id]}
                                            source='sourceName' />
                                    }
                                    subheader={<TimeAgo date={data[id].publishDate} />}
                                    avatar={
                                        <img
                                            src={`${Constants.API_ENDPOINT}/images/${Constants.SOURCE_IMAGES[data[id].sourceName]}.png`}
                                            width={Constants.AVATAR_SIZE}
                                            height={Constants.AVATAR_SIZE}
                                            title={data[id].sourceName}
                                            alt={data[id].sourceName} />
                                    } />
                                <CardContent className={classes.content}>
                                    <div className={classes.card}>
                                        <div className={classes.details}>
                                            <RichTextField
                                                className={classes.title}
                                                record={data[id]}
                                                source='title' />
                                            <RichTextField
                                                className={classes.description}
                                                record={data[id]}
                                                source='description' />
                                        </div>
                                        {data[id].images && data[id].images.length > 0 && (
                                            <Tooltip title={data[id].images[0].description ? StringUtils.decode(data[id].images[0].description) : ''}>
                                                <CardMedia
                                                    className={classes.media}
                                                    image={data[id].images[0].url} />
                                            </Tooltip>
                                        )}
                                    </div>
                                </CardContent>
                                <CardActions className={classes.actions}>
                                    <ShowButton
                                        basePath={basePath}
                                        label='labels.read_more'
                                        record={data[id]} />
                                </CardActions>
                            </CardActionArea>
                        </Card>
                    </Grid>
                ))}
            </Grid>
        </div>
    );
};

ItemGrid.propTypes = {
    basePath : PropTypes.string,
    ids      : PropTypes.arrayOf(PropTypes.number),
    data     : PropTypes.object,
};
