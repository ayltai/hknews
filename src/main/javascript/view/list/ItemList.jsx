import React from 'react';
import { List, Pagination, } from 'react-admin';

import { Constants, } from '../../Constants';
import { ScrollToTopOnMount, } from '../ScrollToTopOnMount';
import { ItemFilter, } from './ItemFilter';
import { ItemGrid, } from './ItemGrid';

export const ItemList = props => (
    <React.Fragment>
        <ScrollToTopOnMount />
        <List
            bulkActionButtons={false}
            exporter={false}
            filters={<ItemFilter />}
            perPage={Constants.NEWS_PER_PAGE}
            pagination={<Pagination rowsPerPageOptions={Constants.NEWS_PER_PAGE_OPTIONS} />}
            {...props}>
            <ItemGrid />
        </List>
    </React.Fragment>
);
