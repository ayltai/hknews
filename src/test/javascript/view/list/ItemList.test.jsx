import { createMuiTheme, MuiThemeProvider, } from '@material-ui/core';
import { render, } from 'enzyme';
import React from 'react';
import { TestContext, } from 'react-admin';
import renderer from 'react-test-renderer';

import { ItemList, } from '../../../../main/javascript/view/list/ItemList';

const component = (
    <TestContext
        initialState={{
            admin     : {
                resources : {
                    sources : {
                        data : {
                            id1 : {
                                title       : '',
                                description : '',
                                images      : [],
                                publishDate : '',
                                sourceName  : '',
                            },
                        },
                        list : {
                            ids         : [
                                'id1',
                            ],
                            params      : {},
                            selectedIds : [],
                            total       : 0,
                        },
                    },
                },
            },
        }}>
        <MuiThemeProvider theme={createMuiTheme({
            palette : {
                type : 'dark',
            },
        })}>
            <ItemList
                basePath='/'
                ids={[]}
                location={{
                    pathname : '',
                }}
                resource='source'>
                <div />
            </ItemList>
        </MuiThemeProvider>
    </TestContext>
);

describe('<ItemList />', () => {
    it('renders without errors', () => {
        expect(renderer.create(component).toJSON()).toMatchSnapshot();
    });

    it('mounts without errors', () => {
        render(component);
    });
});
