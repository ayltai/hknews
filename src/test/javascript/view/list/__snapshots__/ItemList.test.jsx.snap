// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemList /> renders without errors 1`] = `"<div class=\\"list-page RaList-root-1\\" filter=\\"[object Object]\\"><div class=\\"MuiToolbar-root MuiToolbar-regular RaListToolbar-toolbar-7 MuiToolbar-gutters\\"><form class=\\"RaFilter-form-10 RaFilterForm-form-11\\" __versions=\\"[object Object]\\"><div data-source=\\"source\\" class=\\"filter-field RaFilterFormInput-body-13\\"><div class=\\"MuiFormControl-root MuiFormControl-marginDense\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root RaLabeled-label-17 MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense\\" data-shrink=\\"true\\" for=\\"source\\"><span>labels.publisher</span></label><div class=\\"RaLabeled-value-18\\"></div></div><div class=\\"RaFilterFormInput-spacer-14\\">&nbsp;</div></div><div class=\\"RaFilterForm-clearFix-12\\"></div></form><span></span><div class=\\"MuiToolbar-root MuiToolbar-regular RaTopToolbar-root-20 RaListToolbar-actions-8 MuiToolbar-gutters\\"></div></div><div class=\\"RaList-main-2\\"><div class=\\"MuiPaper-root MuiCard-root RaList-content-3 MuiPaper-elevation1 MuiPaper-rounded\\"><div class=\\"makeStyles-container-34\\"><div class=\\"MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2\\"></div></div><span class=\\"MuiTablePagination-root\\" filter=\\"[object Object]\\"><div class=\\"MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters\\"><div class=\\"MuiTablePagination-spacer\\"></div><p class=\\"MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit\\" >ra.navigation.page_rows_per_page</p><div class=\\"MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot\\"><div class=\\"MuiSelect-root MuiSelect-select MuiTablePagination-select MuiSelect-selectMenu MuiInputBase-input\\" tabindex=\\"0\\" role=\\"button\\" aria-haspopup=\\"listbox\\"  >10</div><input aria-hidden=\\"true\\" tabindex=\\"-1\\" class=\\"MuiSelect-nativeInput\\" value=\\"10\\"><svg class=\\"MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M7 10l5 5 5-5z\\"></path></svg></div><p class=\\"MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit\\">ra.navigation.page_range_info</p><div class=\\"RaPaginationActions-actions-30\\"></div></div></span></div></div></div>"`;
